encoding/codecv2/codecv2.go

```diff
@@ -0,0 +1,618 @@
+	if err1 != nil || err2 != nil {
+		log.Crit("Failed to initialize abi types", "err1", err1, "err2", err2)
+	}
+
+	BlobDataProofArgs = abi.Arguments{
+		{Type: bytes32Type, Name: "z"},
+		{Type: bytes32Type, Name: "y"},
+		{Type: bytes48Type, Name: "commitment"},
+		{Type: bytes48Type, Name: "proof"},
+	}
+}
+
+// CodecV2Version denotes the version of the codec.
+const CodecV2Version = 2
+
+// DABlock represents a Data Availability Block.
+type DABlock struct {
+	BlockNumber     uint64
+	Timestamp       uint64
+	BaseFee         *big.Int
+	GasLimit        uint64
+	NumTransactions uint16
+	NumL1Messages   uint16
+}
+
+// DAChunk groups consecutive DABlocks with their transactions.
+type DAChunk struct {
+	Blocks       []*DABlock
+	Transactions [][]*types.TransactionData
+}
+
+// DABatch contains metadata about a batch of DAChunks.
+type DABatch struct {
+	// header
+	Version                uint8
+	BatchIndex             uint64
+	L1MessagePopped        uint64
+	TotalL1MessagePopped   uint64
+	DataHash               common.Hash
+	BlobVersionedHash      common.Hash
+	ParentBatchHash        common.Hash
+	SkippedL1MessageBitmap []byte
+
+	// blob payload
+	blob *kzg4844.Blob
+	z    *kzg4844.Point
+}
+
+// NewDABlock creates a new DABlock from the given encoding.Block and the total number of L1 messages popped before.
+func NewDABlock(block *encoding.Block, totalL1MessagePoppedBefore uint64) (*DABlock, error) {
+	if !block.Header.Number.IsUint64() {
+		return nil, errors.New("block number is not uint64")
+	}
+
+	// note: numL1Messages includes skipped messages
+	numL1Messages := block.NumL1Messages(totalL1MessagePoppedBefore)
+	if numL1Messages > math.MaxUint16 {
+		return nil, errors.New("number of L1 messages exceeds max uint16")
+	}
+
+	// note: numTransactions includes skipped messages
+	numL2Transactions := block.NumL2Transactions()
+	numTransactions := numL1Messages + numL2Transactions
+	if numTransactions > math.MaxUint16 {
+		return nil, errors.New("number of transactions exceeds max uint16")
+	}
+
+	daBlock := DABlock{
+		BlockNumber:     block.Header.Number.Uint64(),
+		Timestamp:       block.Header.Time,
+		BaseFee:         block.Header.BaseFee,
+		GasLimit:        block.Header.GasLimit,
+		NumTransactions: uint16(numTransactions),
+		NumL1Messages:   uint16(numL1Messages),
+	}
+
+	return &daBlock, nil
+}
+
+// Encode serializes the DABlock into a slice of bytes.
+func (b *DABlock) Encode() []byte {
+	bytes := make([]byte, 60)
+	binary.BigEndian.PutUint64(bytes[0:], b.BlockNumber)
+	binary.BigEndian.PutUint64(bytes[8:], b.Timestamp)
+	if b.BaseFee != nil {
+		binary.BigEndian.PutUint64(bytes[40:], b.BaseFee.Uint64())
+	}
+	binary.BigEndian.PutUint64(bytes[48:], b.GasLimit)
+	binary.BigEndian.PutUint16(bytes[56:], b.NumTransactions)
+	binary.BigEndian.PutUint16(bytes[58:], b.NumL1Messages)
+	return bytes
+}
+
+// DecodeDABlock takes a byte slice and decodes it into a DABlock.
+func DecodeDABlock(bytes []byte) (*DABlock, error) {
```

Comment:
why `DecodeDABlock` don't have the `* DABlock` receiver?

Reply:
The only place that uses it is l2geth rollup-sync-service:
```
daBlock, err := codecv0.DecodeDABlock(chunk[startIdx:endIdx])
if err != nil {
	return nil, err
}
```
adding a receiver is also OK. the difference is that the invoke would change to:
```
daBlock := &codecv0.DABlock{}
err := daBlock.DecodeDABlock(chunk[startIdx:endIdx])
if err != nil {
	return nil, err
}
```
so I think we can keep the function as it is.

Reply:
I prefer to keep consistency, because `da-codec` is a public pkg.

Reply:
changed in [d6ff1c1](https://github.com/scroll-tech/da-codec/pull/2/commits/d6ff1c12eca2760e8e909c1a26636b464ec69979).